# HG changeset patch
# User Stelios Voutsinas <stv@roe.ac.uk>
# Date 1417637236 -7200
#      Wed Dec 03 22:07:16 2014 +0200
# Branch 1.11.28-stv-cds-adql-2
# Node ID 5e5206c8b3fc77227d5b17503839e4f982388225
# Parent  65b854cbf712ebcd8d75cba41d324e2538adead2
% Operator support [CDS-Patch]

diff -r 65b854cbf712 -r 5e5206c8b3fc adql/parser/ADQLParser.java
--- a/adql/parser/ADQLParser.java	Wed Dec 03 20:54:03 2014 +0200
+++ b/adql/parser/ADQLParser.java	Wed Dec 03 22:07:16 2014 +0200
@@ -2930,18 +2930,6 @@
     finally { jj_save(15, xla); }
   }
 
-  private boolean jj_3R_163() {
-    Token xsp;
-    xsp = jj_scanpos;
-    if (jj_3R_185()) jj_scanpos = xsp;
-    if (jj_3R_186()) return true;
-    while (true) {
-      xsp = jj_scanpos;
-      if (jj_3R_187()) { jj_scanpos = xsp; break; }
-    }
-    return false;
-  }
-
   private boolean jj_3R_148() {
     if (jj_scan_token(TOP)) return true;
     if (jj_scan_token(UNSIGNED_INTEGER)) return true;
@@ -3280,7 +3268,7 @@
   private boolean jj_3_5() {
     Token xsp;
     xsp = jj_scanpos;
-    if (jj_scan_token(66)) {
+    if (jj_scan_token(67)) {
     jj_scanpos = xsp;
     if (jj_3R_20()) return true;
     }
@@ -3745,7 +3733,7 @@
     }
     }
     xsp = jj_scanpos;
-    if (jj_scan_token(28)) jj_scanpos = xsp;
+    if (jj_scan_token(29)) jj_scanpos = xsp;
     return false;
   }
 
@@ -3794,7 +3782,7 @@
   private boolean jj_3R_49() {
     Token xsp;
     xsp = jj_scanpos;
-    if (jj_scan_token(27)) {
+    if (jj_scan_token(28)) {
     jj_scanpos = xsp;
     if (jj_3R_70()) return true;
     }
@@ -3822,11 +3810,11 @@
   private boolean jj_3R_142() {
     Token xsp;
     xsp = jj_scanpos;
-    if (jj_scan_token(102)) {
-    jj_scanpos = xsp;
     if (jj_scan_token(103)) {
     jj_scanpos = xsp;
-    if (jj_scan_token(104)) return true;
+    if (jj_scan_token(104)) {
+    jj_scanpos = xsp;
+    if (jj_scan_token(105)) return true;
     }
     }
     return false;
@@ -3880,7 +3868,7 @@
     }
     }
     xsp = jj_scanpos;
-    if (jj_scan_token(28)) jj_scanpos = xsp;
+    if (jj_scan_token(29)) jj_scanpos = xsp;
     return false;
   }
 
@@ -3925,7 +3913,7 @@
   private boolean jj_3R_50() {
     Token xsp;
     xsp = jj_scanpos;
-    if (jj_scan_token(27)) {
+    if (jj_scan_token(28)) {
     jj_scanpos = xsp;
     if (jj_3R_71()) return true;
     }
@@ -3981,7 +3969,7 @@
   private boolean jj_3R_192() {
     Token xsp;
     xsp = jj_scanpos;
-    if (jj_scan_token(25)) jj_scanpos = xsp;
+    if (jj_scan_token(26)) jj_scanpos = xsp;
     if (jj_3R_14()) return true;
     return false;
   }
@@ -3994,9 +3982,9 @@
   private boolean jj_3R_190() {
     Token xsp;
     xsp = jj_scanpos;
-    if (jj_scan_token(48)) {
+    if (jj_scan_token(49)) {
     jj_scanpos = xsp;
-    if (jj_scan_token(49)) return true;
+    if (jj_scan_token(50)) return true;
     }
     return false;
   }
@@ -4118,7 +4106,7 @@
     if (jj_3R_16()) return true;
     Token xsp;
     xsp = jj_scanpos;
-    if (jj_scan_token(25)) jj_scanpos = xsp;
+    if (jj_scan_token(26)) jj_scanpos = xsp;
     if (jj_3R_14()) return true;
     return false;
   }
@@ -4192,9 +4180,9 @@
   private boolean jj_3R_101() {
     Token xsp;
     xsp = jj_scanpos;
-    if (jj_scan_token(61)) {
+    if (jj_scan_token(62)) {
     jj_scanpos = xsp;
-    if (jj_scan_token(62)) return true;
+    if (jj_scan_token(63)) return true;
     }
     if (jj_scan_token(LEFT_PAR)) return true;
     if (jj_3R_122()) return true;
@@ -4220,7 +4208,7 @@
     xsp = jj_scanpos;
     if (jj_3R_189()) {
     jj_scanpos = xsp;
-    if (jj_scan_token(104)) return true;
+    if (jj_scan_token(105)) return true;
     }
     xsp = jj_scanpos;
     if (jj_3R_190()) jj_scanpos = xsp;
@@ -4249,19 +4237,19 @@
   private boolean jj_3R_160() {
     Token xsp;
     xsp = jj_scanpos;
-    if (jj_scan_token(50)) {
-    jj_scanpos = xsp;
     if (jj_scan_token(51)) {
     jj_scanpos = xsp;
     if (jj_scan_token(52)) {
     jj_scanpos = xsp;
-    if (jj_scan_token(53)) return true;
+    if (jj_scan_token(53)) {
+    jj_scanpos = xsp;
+    if (jj_scan_token(54)) return true;
     }
     }
     }
     if (jj_scan_token(LEFT_PAR)) return true;
     xsp = jj_scanpos;
-    if (jj_scan_token(22)) jj_scanpos = xsp;
+    if (jj_scan_token(23)) jj_scanpos = xsp;
     if (jj_3R_46()) return true;
     if (jj_scan_token(RIGHT_PAR)) return true;
     return false;
@@ -4279,7 +4267,7 @@
     if (jj_scan_token(LEFT_PAR)) return true;
     Token xsp;
     xsp = jj_scanpos;
-    if (jj_scan_token(22)) jj_scanpos = xsp;
+    if (jj_scan_token(23)) jj_scanpos = xsp;
     xsp = jj_scanpos;
     if (jj_scan_token(10)) {
     jj_scanpos = xsp;
@@ -4299,7 +4287,7 @@
     xsp = jj_scanpos;
     if (jj_3R_188()) {
     jj_scanpos = xsp;
-    if (jj_scan_token(104)) return true;
+    if (jj_scan_token(105)) return true;
     }
     return false;
   }
@@ -4365,7 +4353,7 @@
   private boolean jj_3R_107() {
     Token xsp;
     xsp = jj_scanpos;
-    if (jj_scan_token(38)) jj_scanpos = xsp;
+    if (jj_scan_token(39)) jj_scanpos = xsp;
     if (jj_scan_token(IN)) return true;
     xsp = jj_scanpos;
     if (jj_3_16()) {
@@ -4398,7 +4386,7 @@
   private boolean jj_3R_26() {
     Token xsp;
     xsp = jj_scanpos;
-    if (jj_scan_token(38)) jj_scanpos = xsp;
+    if (jj_scan_token(39)) jj_scanpos = xsp;
     if (jj_scan_token(BETWEEN)) return true;
     if (jj_3R_46()) return true;
     if (jj_scan_token(AND)) return true;
@@ -4430,7 +4418,7 @@
     if (jj_3R_27()) return true;
     Token xsp;
     xsp = jj_scanpos;
-    if (jj_scan_token(38)) jj_scanpos = xsp;
+    if (jj_scan_token(39)) jj_scanpos = xsp;
     if (jj_scan_token(LIKE)) return true;
     return false;
   }
@@ -4455,8 +4443,6 @@
   private boolean jj_3R_106() {
     Token xsp;
     xsp = jj_scanpos;
-    if (jj_scan_token(15)) {
-    jj_scanpos = xsp;
     if (jj_scan_token(16)) {
     jj_scanpos = xsp;
     if (jj_scan_token(17)) {
@@ -4465,7 +4451,9 @@
     jj_scanpos = xsp;
     if (jj_scan_token(19)) {
     jj_scanpos = xsp;
-    if (jj_scan_token(20)) return true;
+    if (jj_scan_token(20)) {
+    jj_scanpos = xsp;
+    if (jj_scan_token(21)) return true;
     }
     }
     }
@@ -4510,7 +4498,7 @@
     if (jj_3R_27()) return true;
     Token xsp;
     xsp = jj_scanpos;
-    if (jj_scan_token(38)) jj_scanpos = xsp;
+    if (jj_scan_token(39)) jj_scanpos = xsp;
     if (jj_scan_token(LIKE)) return true;
     if (jj_3R_27()) return true;
     return false;
@@ -4527,7 +4515,7 @@
     if (jj_scan_token(IS)) return true;
     Token xsp;
     xsp = jj_scanpos;
-    if (jj_scan_token(38)) jj_scanpos = xsp;
+    if (jj_scan_token(39)) jj_scanpos = xsp;
     if (jj_scan_token(NULL)) return true;
     return false;
   }
@@ -4652,9 +4640,9 @@
   private boolean jj_3R_187() {
     Token xsp;
     xsp = jj_scanpos;
-    if (jj_scan_token(36)) {
+    if (jj_scan_token(37)) {
     jj_scanpos = xsp;
-    if (jj_scan_token(37)) return true;
+    if (jj_scan_token(38)) return true;
     }
     xsp = jj_scanpos;
     if (jj_3R_196()) jj_scanpos = xsp;
@@ -4712,6 +4700,18 @@
     return false;
   }
 
+  private boolean jj_3R_163() {
+    Token xsp;
+    xsp = jj_scanpos;
+    if (jj_3R_185()) jj_scanpos = xsp;
+    if (jj_3R_186()) return true;
+    while (true) {
+      xsp = jj_scanpos;
+      if (jj_3R_187()) { jj_scanpos = xsp; break; }
+    }
+    return false;
+  }
+
   /** Generated Token Manager. */
   public ADQLParserTokenManager token_source;
   SimpleCharStream jj_input_stream;
@@ -4735,16 +4735,16 @@
       jj_la1_init_3();
    }
    private static void jj_la1_init_0() {
-      jj_la1_0 = new int[] {0x41,0x0,0x0,0x0,0x0,0x400000,0x800000,0x20,0x0,0x0,0x2000000,0x400,0x304,0x20,0x20,0x20,0x0,0x10,0x10,0x10,0x0,0x0,0x0,0x0,0x2000000,0x2000000,0x2000000,0x0,0x4,0xec000000,0xe0000000,0x10000000,0xe8000000,0xe8000000,0xe0000000,0x10000000,0xe8000000,0xe8000000,0x20,0x0,0xec000000,0x0,0x0,0x0,0x300,0x300,0x4,0x4,0x0,0x304,0x300,0x300,0xc00,0xc00,0x300,0x300,0x4,0x80,0x0,0x4,0x0,0x0,0x0,0x0,0x0,0x4,0x0,0x0,0x1f8000,0x0,0x0,0x304,0x1f8000,0x0,0x0,0x20,0x4,0x400000,0x704,0x0,0x400000,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x20,0x0,0x0,0x304,0x20,0x20,0x0,0x0,0x20,0x304,};
+      jj_la1_0 = new int[] {0x41,0x0,0x0,0x0,0x0,0x800000,0x1000000,0x20,0x0,0x0,0x4000000,0x400,0x304,0x20,0x20,0x20,0x0,0x10,0x10,0x10,0x0,0x0,0x0,0x0,0x4000000,0x4000000,0x4000000,0x0,0x4,0xd8000000,0xc0000000,0x20000000,0xd0000000,0xd0000000,0xc0000000,0x20000000,0xd0000000,0xd0000000,0x20,0x0,0xd8000000,0x0,0x0,0x0,0x300,0x300,0x4,0x4,0x0,0x304,0x300,0x300,0xc00,0xc00,0x300,0x300,0x4,0x80,0x0,0x4,0x0,0x0,0x0,0x0,0x0,0x4,0x0,0x0,0x3f0000,0x0,0x0,0x304,0x3f0000,0x0,0x0,0x20,0x4,0x800000,0x704,0x0,0x800000,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x20,0x0,0x0,0x304,0x20,0x20,0x0,0x0,0x20,0x304,};
    }
    private static void jj_la1_init_1() {
-      jj_la1_1 = new int[] {0x0,0x8,0x2000,0x4000,0x8000,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0xfffc0000,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x30000,0x30000,0x0,0x0,0x0,0x0,0x0,0x1,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x6,0x1,0x0,0x0,0x0,0x0,0x0,0x7c0000,0x0,0x1f800000,0xe07c0000,0x0,0x0,0x0,0x0,0x0,0x0,0x7c0000,0x0,0x0,0x0,0x1f800000,0x40,0x30,0x30,0x40,0x0,0x40,0x40,0x0,0x840,0x1000,0xfffc0000,0x0,0x40,0x40,0x0,0x0,0x0,0xfffc0000,0x3c0000,0x0,0x7c0000,0x60000000,0x4000000,0x4000000,0x4000000,0x4000000,0xe0000000,0x0,0x1f800000,0xe0000000,0xe07c0000,0x0,0x0,0x0,0x0,0x0,0xfffc0000,};
+      jj_la1_1 = new int[] {0x0,0x10,0x4000,0x8000,0x10000,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0xfff80000,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x60000,0x60000,0x0,0x0,0x0,0x0,0x0,0x3,0x1,0x0,0x1,0x1,0x1,0x0,0x1,0x1,0x0,0xc,0x3,0x0,0x0,0x0,0x0,0x0,0xf80000,0x0,0x3f000000,0xc0f80000,0x0,0x0,0x0,0x0,0x0,0x0,0xf80000,0x0,0x0,0x0,0x3f000000,0x80,0x60,0x60,0x80,0x0,0x80,0x80,0x0,0x1080,0x2000,0xfff80000,0x0,0x80,0x80,0x0,0x0,0x0,0xfff80000,0x780000,0x0,0xf80000,0xc0000000,0x8000000,0x8000000,0x8000000,0x8000000,0xc0000000,0x0,0x3f000000,0xc0000000,0xc0f80000,0x0,0x0,0x0,0x0,0x0,0xfff80000,};
    }
    private static void jj_la1_init_2() {
-      jj_la1_2 = new int[] {0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x7ffffff,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x7fffffb,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x4,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x7ffffff,0x0,0x0,0x0,0x0,0x0,0x0,0x7ffffff,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0xb,0x0,0x0,0x7fffffb,0x7fffffb,0x0,0x0,0x7fff0,0x7f80000,0x0,0x7ffffff,};
+      jj_la1_2 = new int[] {0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0xfffffff,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0xffffff7,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x8,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0xfffffff,0x0,0x0,0x0,0x0,0x0,0x0,0xfffffff,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x17,0x0,0x0,0xffffff7,0xffffff7,0x0,0x0,0xfffe0,0xff00000,0x0,0xfffffff,};
    }
    private static void jj_la1_init_3() {
-      jj_la1_3 = new int[] {0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x18,0x18,0x0,0x0,0x1d9,0x0,0x0,0x0,0x18,0x0,0x0,0x0,0x118,0x118,0x0,0x0,0x0,0x18,0x0,0x18,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x1,0x1c0,0x180,0x0,0x0,0x1d8,0x19,0x0,0x1d8,0x0,0x0,0x0,0x0,0x0,0x0,0x1d8,0x0,0x0,0x19,0x18,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x1d9,0x0,0x0,0x0,0x0,0x0,0x0,0x1d9,0x0,0x0,0x0,0x0,0x18,0x18,0x18,0x18,0x0,0x0,0x0,0x10,0x1d8,0x0,0x0,0x0,0x0,0x0,0x1d9,};
+      jj_la1_3 = new int[] {0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x30,0x30,0x0,0x0,0x3b2,0x0,0x0,0x0,0x30,0x0,0x0,0x0,0x230,0x230,0x0,0x0,0x0,0x30,0x0,0x30,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x2,0x380,0x300,0x0,0x0,0x3b0,0x32,0x0,0x3b0,0x0,0x0,0x0,0x0,0x0,0x0,0x3b0,0x0,0x0,0x32,0x30,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x3b2,0x0,0x0,0x0,0x0,0x0,0x0,0x3b2,0x0,0x0,0x0,0x0,0x30,0x30,0x30,0x30,0x0,0x0,0x0,0x20,0x3b0,0x0,0x0,0x0,0x0,0x0,0x3b2,};
    }
   final private JJCalls[] jj_2_rtns = new JJCalls[16];
   private boolean jj_rescan = false;
@@ -4933,7 +4933,7 @@
   /** Generate ParseException. */
   public ParseException generateParseException() {
     jj_expentries.clear();
-    boolean[] la1tokens = new boolean[106];
+    boolean[] la1tokens = new boolean[107];
     if (jj_kind >= 0) {
       la1tokens[jj_kind] = true;
       jj_kind = -1;
@@ -4956,7 +4956,7 @@
         }
       }
     }
-    for (int i = 0; i < 106; i++) {
+    for (int i = 0; i < 107; i++) {
       if (la1tokens[i]) {
         jj_expentry = new int[1];
         jj_expentry[0] = i;
diff -r 65b854cbf712 -r 5e5206c8b3fc adql/parser/ADQLParserConstants.java
--- a/adql/parser/ADQLParserConstants.java	Wed Dec 03 20:54:03 2014 +0200
+++ b/adql/parser/ADQLParserConstants.java	Wed Dec 03 22:07:16 2014 +0200
@@ -31,179 +31,181 @@
   /** RegularExpression Id. */
   int DIVIDE = 11;
   /** RegularExpression Id. */
-  int BIT_AND = 12;
+  int MODULO = 12;
   /** RegularExpression Id. */
-  int BIT_OR = 13;
+  int BIT_AND = 13;
   /** RegularExpression Id. */
-  int BIT_XOR = 14;
+  int BIT_OR = 14;
   /** RegularExpression Id. */
-  int EQUAL = 15;
+  int BIT_XOR = 15;
   /** RegularExpression Id. */
-  int NOT_EQUAL = 16;
+  int EQUAL = 16;
   /** RegularExpression Id. */
-  int LESS_THAN = 17;
+  int NOT_EQUAL = 17;
   /** RegularExpression Id. */
-  int LESS_EQUAL_THAN = 18;
+  int LESS_THAN = 18;
   /** RegularExpression Id. */
-  int GREATER_THAN = 19;
+  int LESS_EQUAL_THAN = 19;
   /** RegularExpression Id. */
-  int GREATER_EQUAL_THAN = 20;
+  int GREATER_THAN = 20;
   /** RegularExpression Id. */
-  int SELECT = 21;
+  int GREATER_EQUAL_THAN = 21;
   /** RegularExpression Id. */
-  int QUANTIFIER = 22;
+  int SELECT = 22;
   /** RegularExpression Id. */
-  int TOP = 23;
+  int QUANTIFIER = 23;
   /** RegularExpression Id. */
-  int FROM = 24;
+  int TOP = 24;
   /** RegularExpression Id. */
-  int AS = 25;
+  int FROM = 25;
   /** RegularExpression Id. */
-  int NATURAL = 26;
+  int AS = 26;
   /** RegularExpression Id. */
-  int INNER = 27;
+  int NATURAL = 27;
   /** RegularExpression Id. */
-  int OUTER = 28;
+  int INNER = 28;
   /** RegularExpression Id. */
-  int RIGHT = 29;
+  int OUTER = 29;
   /** RegularExpression Id. */
-  int LEFT = 30;
+  int RIGHT = 30;
   /** RegularExpression Id. */
-  int FULL = 31;
+  int LEFT = 31;
   /** RegularExpression Id. */
-  int JOIN = 32;
+  int FULL = 32;
   /** RegularExpression Id. */
-  int ON = 33;
+  int JOIN = 33;
   /** RegularExpression Id. */
-  int USING = 34;
+  int ON = 34;
   /** RegularExpression Id. */
-  int WHERE = 35;
+  int USING = 35;
   /** RegularExpression Id. */
-  int AND = 36;
+  int WHERE = 36;
   /** RegularExpression Id. */
-  int OR = 37;
+  int AND = 37;
   /** RegularExpression Id. */
-  int NOT = 38;
+  int OR = 38;
   /** RegularExpression Id. */
-  int IS = 39;
+  int NOT = 39;
   /** RegularExpression Id. */
-  int NULL = 40;
+  int IS = 40;
   /** RegularExpression Id. */
-  int BETWEEN = 41;
+  int NULL = 41;
   /** RegularExpression Id. */
-  int LIKE = 42;
+  int BETWEEN = 42;
   /** RegularExpression Id. */
-  int IN = 43;
+  int LIKE = 43;
   /** RegularExpression Id. */
-  int EXISTS = 44;
+  int IN = 44;
   /** RegularExpression Id. */
-  int GROUP_BY = 45;
+  int EXISTS = 45;
   /** RegularExpression Id. */
-  int HAVING = 46;
+  int GROUP_BY = 46;
   /** RegularExpression Id. */
-  int ORDER_BY = 47;
+  int HAVING = 47;
   /** RegularExpression Id. */
-  int ASC = 48;
+  int ORDER_BY = 48;
   /** RegularExpression Id. */
-  int DESC = 49;
+  int ASC = 49;
   /** RegularExpression Id. */
-  int AVG = 50;
+  int DESC = 50;
   /** RegularExpression Id. */
-  int MAX = 51;
+  int AVG = 51;
   /** RegularExpression Id. */
-  int MIN = 52;
+  int MAX = 52;
   /** RegularExpression Id. */
-  int SUM = 53;
+  int MIN = 53;
   /** RegularExpression Id. */
-  int COUNT = 54;
+  int SUM = 54;
   /** RegularExpression Id. */
-  int BOX = 55;
+  int COUNT = 55;
   /** RegularExpression Id. */
-  int CENTROID = 56;
+  int BOX = 56;
   /** RegularExpression Id. */
-  int CIRCLE = 57;
+  int CENTROID = 57;
   /** RegularExpression Id. */
-  int POINT = 58;
+  int CIRCLE = 58;
   /** RegularExpression Id. */
-  int POLYGON = 59;
+  int POINT = 59;
   /** RegularExpression Id. */
-  int REGION = 60;
+  int POLYGON = 60;
   /** RegularExpression Id. */
-  int CONTAINS = 61;
+  int REGION = 61;
   /** RegularExpression Id. */
-  int INTERSECTS = 62;
+  int CONTAINS = 62;
   /** RegularExpression Id. */
-  int AREA = 63;
+  int INTERSECTS = 63;
   /** RegularExpression Id. */
-  int COORD1 = 64;
+  int AREA = 64;
   /** RegularExpression Id. */
-  int COORD2 = 65;
+  int COORD1 = 65;
   /** RegularExpression Id. */
-  int COORDSYS = 66;
+  int COORD2 = 66;
   /** RegularExpression Id. */
-  int DISTANCE = 67;
+  int COORDSYS = 67;
   /** RegularExpression Id. */
-  int ABS = 68;
+  int DISTANCE = 68;
   /** RegularExpression Id. */
-  int CEILING = 69;
+  int ABS = 69;
   /** RegularExpression Id. */
-  int DEGREES = 70;
+  int CEILING = 70;
   /** RegularExpression Id. */
-  int EXP = 71;
+  int DEGREES = 71;
   /** RegularExpression Id. */
-  int FLOOR = 72;
+  int EXP = 72;
   /** RegularExpression Id. */
-  int LOG = 73;
+  int FLOOR = 73;
   /** RegularExpression Id. */
-  int LOG10 = 74;
+  int LOG = 74;
   /** RegularExpression Id. */
-  int MOD = 75;
+  int LOG10 = 75;
   /** RegularExpression Id. */
-  int PI = 76;
+  int MOD = 76;
   /** RegularExpression Id. */
-  int POWER = 77;
+  int PI = 77;
   /** RegularExpression Id. */
-  int RADIANS = 78;
+  int POWER = 78;
   /** RegularExpression Id. */
-  int RAND = 79;
+  int RADIANS = 79;
   /** RegularExpression Id. */
-  int ROUND = 80;
+  int RAND = 80;
   /** RegularExpression Id. */
-  int SQRT = 81;
+  int ROUND = 81;
   /** RegularExpression Id. */
-  int TRUNCATE = 82;
+  int SQRT = 82;
   /** RegularExpression Id. */
-  int ACOS = 83;
+  int TRUNCATE = 83;
   /** RegularExpression Id. */
-  int ASIN = 84;
+  int ACOS = 84;
   /** RegularExpression Id. */
-  int ATAN = 85;
+  int ASIN = 85;
   /** RegularExpression Id. */
-  int ATAN2 = 86;
+  int ATAN = 86;
   /** RegularExpression Id. */
-  int COS = 87;
+  int ATAN2 = 87;
   /** RegularExpression Id. */
-  int COT = 88;
+  int COS = 88;
   /** RegularExpression Id. */
-  int SIN = 89;
+  int COT = 89;
   /** RegularExpression Id. */
-  int TAN = 90;
+  int SIN = 90;
   /** RegularExpression Id. */
-  int STRING_LITERAL = 96;
+  int TAN = 91;
   /** RegularExpression Id. */
-  int DELIMITED_IDENTIFIER = 99;
+  int STRING_LITERAL = 97;
   /** RegularExpression Id. */
-  int REGULAR_IDENTIFIER = 100;
+  int DELIMITED_IDENTIFIER = 100;
   /** RegularExpression Id. */
-  int Letter = 101;
+  int REGULAR_IDENTIFIER = 101;
   /** RegularExpression Id. */
-  int SCIENTIFIC_NUMBER = 102;
+  int Letter = 102;
   /** RegularExpression Id. */
-  int UNSIGNED_FLOAT = 103;
+  int SCIENTIFIC_NUMBER = 103;
   /** RegularExpression Id. */
-  int UNSIGNED_INTEGER = 104;
+  int UNSIGNED_FLOAT = 104;
   /** RegularExpression Id. */
-  int DIGIT = 105;
+  int UNSIGNED_INTEGER = 105;
+  /** RegularExpression Id. */
+  int DIGIT = 106;
 
   /** Lexical state. */
   int DEFAULT = 0;
@@ -228,6 +230,7 @@
     "\"-\"",
     "\"*\"",
     "\"/\"",
+    "\"%\"",
     "\"&\"",
     "\"|\"",
     "\"^\"",
@@ -307,14 +310,14 @@
     "\"COT\"",
     "\"SIN\"",
     "\"TAN\"",
-    "<token of kind 91>",
     "<token of kind 92>",
     "<token of kind 93>",
+    "<token of kind 94>",
     "\"\\\'\"",
-    "<token of kind 95>",
+    "<token of kind 96>",
     "\"\\\'\"",
     "\"\\\"\"",
-    "<token of kind 98>",
+    "<token of kind 99>",
     "\"\\\"\"",
     "<REGULAR_IDENTIFIER>",
     "<Letter>",
diff -r 65b854cbf712 -r 5e5206c8b3fc adql/parser/ADQLParserTokenManager.java
--- a/adql/parser/ADQLParserTokenManager.java	Wed Dec 03 20:54:03 2014 +0200
+++ b/adql/parser/ADQLParserTokenManager.java	Wed Dec 03 22:07:16 2014 +0200
@@ -50,7 +50,7 @@
    switch(curChar)
    {
       case 34:
-         return jjStartNfaWithStates_3(0, 99, 1);
+         return jjStartNfaWithStates_3(0, 100, 1);
       default :
          return jjMoveNfa_3(0, 0);
    }
@@ -87,15 +87,15 @@
                case 0:
                   if ((0xfffffffbffffffffL & l) != 0L)
                   {
-                     if (kind > 98)
-                        kind = 98;
+                     if (kind > 99)
+                        kind = 99;
                   }
                   else if (curChar == 34)
                      jjstateSet[jjnewStateCnt++] = 1;
                   break;
                case 1:
-                  if (curChar == 34 && kind > 98)
-                     kind = 98;
+                  if (curChar == 34 && kind > 99)
+                     kind = 99;
                   break;
                case 2:
                   if (curChar == 34)
@@ -113,7 +113,7 @@
             switch(jjstateSet[--i])
             {
                case 0:
-                  kind = 98;
+                  kind = 99;
                   break;
                default : break;
             }
@@ -128,8 +128,8 @@
             switch(jjstateSet[--i])
             {
                case 0:
-                  if ((jjbitVec0[i2] & l2) != 0L && kind > 98)
-                     kind = 98;
+                  if ((jjbitVec0[i2] & l2) != 0L && kind > 99)
+                     kind = 99;
                   break;
                default : break;
             }
@@ -153,180 +153,180 @@
    switch (pos)
    {
       case 0:
-         if ((active0 & 0x10L) != 0L)
-            return 37;
-         if ((active0 & 0x8005001002000000L) != 0L || (active1 & 0x780010L) != 0L)
+         if ((active0 & 0x4000000000000L) != 0L || (active1 & 0x90L) != 0L)
          {
-            jjmatchedKind = 100;
-            return 16;
-         }
-         if ((active0 & 0x60000L) != 0L)
-            return 3;
-         if ((active0 & 0x2000000000000L) != 0L || (active1 & 0x48L) != 0L)
-         {
-            jjmatchedKind = 100;
+            jjmatchedKind = 101;
             return 13;
          }
-         if ((active0 & 0x7ff8ffeffda00000L) != 0L || (active1 & 0x787ffa7L) != 0L)
+         if ((active0 & 0xfff1ffdffb400000L) != 0L || (active1 & 0xf0fff4eL) != 0L)
          {
-            jjmatchedKind = 100;
+            jjmatchedKind = 101;
+            return 37;
+         }
+         if ((active0 & 0x10L) != 0L)
             return 38;
+         if ((active0 & 0xc0000L) != 0L)
+            return 3;
+         if ((active0 & 0xa002004000000L) != 0L || (active1 & 0xf00021L) != 0L)
+         {
+            jjmatchedKind = 101;
+            return 16;
          }
          if ((active0 & 0x200L) != 0L)
             return 19;
          return -1;
       case 1:
-         if ((active1 & 0x8L) != 0L)
+         if ((active1 & 0x10L) != 0L)
          {
             if (jjmatchedPos != 1)
             {
-               jjmatchedKind = 100;
+               jjmatchedKind = 101;
                jjmatchedPos = 1;
             }
             return 12;
          }
-         if ((active0 & 0x400188a20a000000L) != 0L || (active1 & 0x101000L) != 0L)
-            return 38;
-         if ((active0 & 0xbffe775df5a00000L) != 0L || (active1 & 0x7efeff7L) != 0L)
+         if ((active0 & 0x8003114414000000L) != 0L || (active1 & 0x202000L) != 0L)
+            return 37;
+         if ((active0 & 0x7ffceebbeb400000L) != 0L || (active1 & 0xfdfdfefL) != 0L)
          {
             if (jjmatchedPos != 1)
             {
-               jjmatchedKind = 100;
+               jjmatchedKind = 101;
                jjmatchedPos = 1;
             }
-            return 38;
+            return 37;
          }
          return -1;
       case 2:
-         if ((active0 & 0xff42f70dfd200000L) != 0L || (active1 & 0x7fe167L) != 0L)
+         if ((active0 & 0xfe85ee1bfa400000L) != 0L || (active1 & 0xffc2cfL) != 0L)
          {
             if (jjmatchedPos != 2)
             {
-               jjmatchedKind = 100;
+               jjmatchedKind = 101;
                jjmatchedPos = 2;
             }
-            return 38;
+            return 37;
          }
-         if ((active0 & 0xbd005000800000L) != 0L || (active1 & 0x7800e90L) != 0L)
-            return 38;
-         if ((active1 & 0x8L) != 0L)
+         if ((active0 & 0x17a00a001000000L) != 0L || (active1 & 0xf001d20L) != 0L)
+            return 37;
+         if ((active1 & 0x10L) != 0L)
          {
             if (jjmatchedPos != 2)
             {
-               jjmatchedKind = 100;
+               jjmatchedKind = 101;
                jjmatchedPos = 2;
             }
             return 11;
          }
          return -1;
       case 3:
-         if ((active1 & 0x400L) != 0L)
+         if ((active0 & 0xfe81e41878400000L) != 0L || (active1 & 0xac2ceL) != 0L)
          {
             if (jjmatchedPos != 3)
             {
-               jjmatchedKind = 100;
+               jjmatchedKind = 101;
+               jjmatchedPos = 3;
+            }
+            return 37;
+         }
+         if ((active1 & 0x800L) != 0L)
+         {
+            if (jjmatchedPos != 3)
+            {
+               jjmatchedKind = 101;
                jjmatchedPos = 3;
             }
             return 21;
          }
-         if ((active0 & 0x80020501c1000000L) != 0L || (active1 & 0x7a8000L) != 0L)
-            return 38;
-         if ((active0 & 0x7f40f20c3c200000L) != 0L || (active1 & 0x56167L) != 0L)
+         if ((active1 & 0x10L) != 0L)
          {
             if (jjmatchedPos != 3)
             {
-               jjmatchedKind = 100;
-               jjmatchedPos = 3;
-            }
-            return 38;
-         }
-         if ((active1 & 0x8L) != 0L)
-         {
-            if (jjmatchedPos != 3)
-            {
-               jjmatchedKind = 100;
+               jjmatchedKind = 101;
                jjmatchedPos = 3;
             }
             return 10;
          }
+         if ((active0 & 0x40a0382000000L) != 0L || (active1 & 0xf50001L) != 0L)
+            return 37;
          return -1;
       case 4:
-         if ((active0 & 0x440000c38000000L) != 0L || (active1 & 0x12100L) != 0L)
-            return 38;
-         if ((active1 & 0x400400L) != 0L)
+         if ((active0 & 0x880001870000000L) != 0L || (active1 & 0x24200L) != 0L)
+            return 37;
+         if ((active1 & 0x800800L) != 0L)
             return 21;
-         if ((active0 & 0x7b00f20004200000L) != 0L || (active1 & 0x4406fL) != 0L)
+         if ((active0 & 0xf601e40008400000L) != 0L || (active1 & 0x880deL) != 0L)
          {
-            jjmatchedKind = 100;
+            jjmatchedKind = 101;
             jjmatchedPos = 4;
-            return 38;
+            return 37;
          }
          return -1;
       case 5:
-         if ((active0 & 0xa00000000000L) != 0L)
+         if ((active0 & 0xd200040008000000L) != 0L || (active1 & 0x880d8L) != 0L)
+         {
+            jjmatchedKind = 101;
+            jjmatchedPos = 5;
+            return 37;
+         }
+         if ((active0 & 0x2400a00000400000L) != 0L)
+            return 37;
+         if ((active1 & 0x6L) != 0L)
+            return 21;
+         if ((active0 & 0x1400000000000L) != 0L)
          {
             if (jjmatchedPos < 4)
             {
-               jjmatchedKind = 100;
+               jjmatchedKind = 101;
                jjmatchedPos = 4;
             }
             return -1;
          }
-         if ((active0 & 0x6900020004000000L) != 0L || (active1 & 0x4406cL) != 0L)
-         {
-            jjmatchedKind = 100;
-            jjmatchedPos = 5;
-            return 38;
-         }
-         if ((active0 & 0x1200500000200000L) != 0L)
-            return 38;
-         if ((active1 & 0x3L) != 0L)
-            return 21;
          return -1;
       case 6:
-         if ((active0 & 0x6100000000000000L) != 0L || (active1 & 0x4000cL) != 0L)
+         if ((active0 & 0xc200000000000000L) != 0L || (active1 & 0x80018L) != 0L)
          {
-            jjmatchedKind = 100;
+            jjmatchedKind = 101;
             jjmatchedPos = 6;
-            return 38;
+            return 37;
          }
-         if ((active0 & 0xa00000000000L) != 0L)
+         if ((active0 & 0x1000040008000000L) != 0L || (active1 & 0x80c0L) != 0L)
+            return 37;
+         if ((active0 & 0x1400000000000L) != 0L)
          {
             if (jjmatchedPos < 4)
             {
-               jjmatchedKind = 100;
+               jjmatchedKind = 101;
                jjmatchedPos = 4;
             }
             return -1;
          }
-         if ((active0 & 0x800020004000000L) != 0L || (active1 & 0x4060L) != 0L)
-            return 38;
          return -1;
       case 7:
-         if ((active0 & 0xa00000000000L) != 0L)
+         if ((active0 & 0x4200000000000000L) != 0L || (active1 & 0x80018L) != 0L)
+            return 37;
+         if ((active0 & 0x8000000000000000L) != 0L)
+         {
+            jjmatchedKind = 101;
+            jjmatchedPos = 7;
+            return 37;
+         }
+         if ((active0 & 0x1400000000000L) != 0L)
          {
             if (jjmatchedPos < 4)
             {
-               jjmatchedKind = 100;
+               jjmatchedKind = 101;
                jjmatchedPos = 4;
             }
             return -1;
          }
-         if ((active0 & 0x2100000000000000L) != 0L || (active1 & 0x4000cL) != 0L)
-            return 38;
-         if ((active0 & 0x4000000000000000L) != 0L)
-         {
-            jjmatchedKind = 100;
-            jjmatchedPos = 7;
-            return 38;
-         }
          return -1;
       case 8:
-         if ((active0 & 0x4000000000000000L) != 0L)
+         if ((active0 & 0x8000000000000000L) != 0L)
          {
-            jjmatchedKind = 100;
+            jjmatchedKind = 101;
             jjmatchedPos = 8;
-            return 38;
+            return 37;
          }
          return -1;
       default :
@@ -342,11 +342,13 @@
    switch(curChar)
    {
       case 34:
-         return jjStopAtPos(0, 97);
+         return jjStopAtPos(0, 98);
+      case 37:
+         return jjStopAtPos(0, 12);
       case 38:
-         return jjStopAtPos(0, 12);
+         return jjStopAtPos(0, 13);
       case 39:
-         return jjStopAtPos(0, 94);
+         return jjStopAtPos(0, 95);
       case 40:
          return jjStopAtPos(0, 2);
       case 41:
@@ -360,83 +362,83 @@
       case 45:
          return jjStartNfaWithStates_0(0, 9, 19);
       case 46:
-         return jjStartNfaWithStates_0(0, 4, 37);
+         return jjStartNfaWithStates_0(0, 4, 38);
       case 47:
          return jjStopAtPos(0, 11);
       case 59:
          return jjStopAtPos(0, 6);
       case 60:
-         jjmatchedKind = 17;
-         return jjMoveStringLiteralDfa1_0(0x40000L, 0x0L);
+         jjmatchedKind = 18;
+         return jjMoveStringLiteralDfa1_0(0x80000L, 0x0L);
       case 61:
+         return jjStopAtPos(0, 16);
+      case 62:
+         jjmatchedKind = 20;
+         return jjMoveStringLiteralDfa1_0(0x200000L, 0x0L);
+      case 94:
          return jjStopAtPos(0, 15);
-      case 62:
-         jjmatchedKind = 19;
-         return jjMoveStringLiteralDfa1_0(0x100000L, 0x0L);
-      case 94:
-         return jjStopAtPos(0, 14);
       case 65:
       case 97:
-         return jjMoveStringLiteralDfa1_0(0x8005001002000000L, 0x780010L);
+         return jjMoveStringLiteralDfa1_0(0xa002004000000L, 0xf00021L);
       case 66:
       case 98:
-         return jjMoveStringLiteralDfa1_0(0x80020000000000L, 0x0L);
+         return jjMoveStringLiteralDfa1_0(0x100040000000000L, 0x0L);
       case 67:
       case 99:
-         return jjMoveStringLiteralDfa1_0(0x2340000000000000L, 0x1800027L);
+         return jjMoveStringLiteralDfa1_0(0x4680000000000000L, 0x300004eL);
       case 68:
       case 100:
-         return jjMoveStringLiteralDfa1_0(0x2000000000000L, 0x48L);
+         return jjMoveStringLiteralDfa1_0(0x4000000000000L, 0x90L);
       case 69:
       case 101:
-         return jjMoveStringLiteralDfa1_0(0x100000000000L, 0x80L);
+         return jjMoveStringLiteralDfa1_0(0x200000000000L, 0x100L);
       case 70:
       case 102:
-         return jjMoveStringLiteralDfa1_0(0x81000000L, 0x100L);
+         return jjMoveStringLiteralDfa1_0(0x102000000L, 0x200L);
       case 71:
       case 103:
-         return jjMoveStringLiteralDfa1_0(0x200000000000L, 0x0L);
+         return jjMoveStringLiteralDfa1_0(0x400000000000L, 0x0L);
       case 72:
       case 104:
-         return jjMoveStringLiteralDfa1_0(0x400000000000L, 0x0L);
+         return jjMoveStringLiteralDfa1_0(0x800000000000L, 0x0L);
       case 73:
       case 105:
-         return jjMoveStringLiteralDfa1_0(0x4000088008000000L, 0x0L);
+         return jjMoveStringLiteralDfa1_0(0x8000110010000000L, 0x0L);
       case 74:
       case 106:
-         return jjMoveStringLiteralDfa1_0(0x100000000L, 0x0L);
+         return jjMoveStringLiteralDfa1_0(0x200000000L, 0x0L);
       case 76:
       case 108:
-         return jjMoveStringLiteralDfa1_0(0x40040000000L, 0x600L);
+         return jjMoveStringLiteralDfa1_0(0x80080000000L, 0xc00L);
       case 77:
       case 109:
-         return jjMoveStringLiteralDfa1_0(0x18000000000000L, 0x800L);
+         return jjMoveStringLiteralDfa1_0(0x30000000000000L, 0x1000L);
       case 78:
       case 110:
-         return jjMoveStringLiteralDfa1_0(0x14004000000L, 0x0L);
+         return jjMoveStringLiteralDfa1_0(0x28008000000L, 0x0L);
       case 79:
       case 111:
-         return jjMoveStringLiteralDfa1_0(0x802210000000L, 0x0L);
+         return jjMoveStringLiteralDfa1_0(0x1004420000000L, 0x0L);
       case 80:
       case 112:
-         return jjMoveStringLiteralDfa1_0(0xc00000000000000L, 0x3000L);
+         return jjMoveStringLiteralDfa1_0(0x1800000000000000L, 0x6000L);
       case 82:
       case 114:
-         return jjMoveStringLiteralDfa1_0(0x1000000020000000L, 0x1c000L);
+         return jjMoveStringLiteralDfa1_0(0x2000000040000000L, 0x38000L);
       case 83:
       case 115:
-         return jjMoveStringLiteralDfa1_0(0x20000000200000L, 0x2020000L);
+         return jjMoveStringLiteralDfa1_0(0x40000000400000L, 0x4040000L);
       case 84:
       case 116:
-         return jjMoveStringLiteralDfa1_0(0x800000L, 0x4040000L);
+         return jjMoveStringLiteralDfa1_0(0x1000000L, 0x8080000L);
       case 85:
       case 117:
-         return jjMoveStringLiteralDfa1_0(0x400000000L, 0x0L);
+         return jjMoveStringLiteralDfa1_0(0x800000000L, 0x0L);
       case 87:
       case 119:
-         return jjMoveStringLiteralDfa1_0(0x800000000L, 0x0L);
+         return jjMoveStringLiteralDfa1_0(0x1000000000L, 0x0L);
       case 124:
-         jjmatchedKind = 13;
+         jjmatchedKind = 14;
          return jjMoveStringLiteralDfa1_0(0x80L, 0x0L);
       default :
          return jjMoveNfa_0(0, 0);
@@ -452,80 +454,80 @@
    switch(curChar)
    {
       case 61:
-         if ((active0 & 0x40000L) != 0L)
-            return jjStopAtPos(1, 18);
-         else if ((active0 & 0x100000L) != 0L)
-            return jjStopAtPos(1, 20);
+         if ((active0 & 0x80000L) != 0L)
+            return jjStopAtPos(1, 19);
+         else if ((active0 & 0x200000L) != 0L)
+            return jjStopAtPos(1, 21);
          break;
       case 65:
       case 97:
-         return jjMoveStringLiteralDfa2_0(active0, 0x8400004000000L, active1, 0x400c000L);
+         return jjMoveStringLiteralDfa2_0(active0, 0x10800008000000L, active1, 0x8018000L);
       case 66:
       case 98:
-         return jjMoveStringLiteralDfa2_0(active0, 0L, active1, 0x10L);
+         return jjMoveStringLiteralDfa2_0(active0, 0L, active1, 0x20L);
       case 67:
       case 99:
-         return jjMoveStringLiteralDfa2_0(active0, 0L, active1, 0x80000L);
+         return jjMoveStringLiteralDfa2_0(active0, 0L, active1, 0x100000L);
       case 69:
       case 101:
-         return jjMoveStringLiteralDfa2_0(active0, 0x1102020040200000L, active1, 0x60L);
+         return jjMoveStringLiteralDfa2_0(active0, 0x2204040080400000L, active1, 0xc0L);
       case 72:
       case 104:
-         return jjMoveStringLiteralDfa2_0(active0, 0x800000000L, active1, 0L);
+         return jjMoveStringLiteralDfa2_0(active0, 0x1000000000L, active1, 0L);
       case 73:
       case 105:
-         if ((active1 & 0x1000L) != 0L)
-            return jjStartNfaWithStates_0(1, 76, 38);
-         return jjMoveStringLiteralDfa2_0(active0, 0x210040020000000L, active1, 0x2000008L);
+         if ((active1 & 0x2000L) != 0L)
+            return jjStartNfaWithStates_0(1, 77, 37);
+         return jjMoveStringLiteralDfa2_0(active0, 0x420080040000000L, active1, 0x4000010L);
       case 76:
       case 108:
-         return jjMoveStringLiteralDfa2_0(active0, 0L, active1, 0x100L);
+         return jjMoveStringLiteralDfa2_0(active0, 0L, active1, 0x200L);
       case 78:
       case 110:
-         if ((active0 & 0x200000000L) != 0L)
-            return jjStartNfaWithStates_0(1, 33, 38);
-         else if ((active0 & 0x80000000000L) != 0L)
+         if ((active0 & 0x400000000L) != 0L)
+            return jjStartNfaWithStates_0(1, 34, 37);
+         else if ((active0 & 0x100000000000L) != 0L)
          {
-            jjmatchedKind = 43;
+            jjmatchedKind = 44;
             jjmatchedPos = 1;
          }
-         return jjMoveStringLiteralDfa2_0(active0, 0x4000001008000000L, active1, 0L);
+         return jjMoveStringLiteralDfa2_0(active0, 0x8000002010000000L, active1, 0L);
       case 79:
       case 111:
-         return jjMoveStringLiteralDfa2_0(active0, 0x2cc0004100800000L, active1, 0x1812e07L);
+         return jjMoveStringLiteralDfa2_0(active0, 0x5980008201000000L, active1, 0x3025c0eL);
       case 81:
       case 113:
-         return jjMoveStringLiteralDfa2_0(active0, 0L, active1, 0x20000L);
+         return jjMoveStringLiteralDfa2_0(active0, 0L, active1, 0x40000L);
       case 82:
       case 114:
-         if ((active0 & 0x2000000000L) != 0L)
+         if ((active0 & 0x4000000000L) != 0L)
          {
-            jjmatchedKind = 37;
+            jjmatchedKind = 38;
             jjmatchedPos = 1;
          }
-         return jjMoveStringLiteralDfa2_0(active0, 0x8000a00001000000L, active1, 0x40000L);
+         return jjMoveStringLiteralDfa2_0(active0, 0x1400002000000L, active1, 0x80001L);
       case 83:
       case 115:
-         if ((active0 & 0x2000000L) != 0L)
+         if ((active0 & 0x4000000L) != 0L)
          {
-            jjmatchedKind = 25;
+            jjmatchedKind = 26;
             jjmatchedPos = 1;
          }
-         else if ((active0 & 0x8000000000L) != 0L)
-            return jjStartNfaWithStates_0(1, 39, 38);
-         return jjMoveStringLiteralDfa2_0(active0, 0x1000400000000L, active1, 0x100000L);
+         else if ((active0 & 0x10000000000L) != 0L)
+            return jjStartNfaWithStates_0(1, 40, 37);
+         return jjMoveStringLiteralDfa2_0(active0, 0x2000800000000L, active1, 0x200000L);
       case 84:
       case 116:
-         return jjMoveStringLiteralDfa2_0(active0, 0L, active1, 0x600000L);
+         return jjMoveStringLiteralDfa2_0(active0, 0L, active1, 0xc00000L);
       case 85:
       case 117:
-         return jjMoveStringLiteralDfa2_0(active0, 0x20010090000000L, active1, 0L);
+         return jjMoveStringLiteralDfa2_0(active0, 0x40020120000000L, active1, 0L);
       case 86:
       case 118:
-         return jjMoveStringLiteralDfa2_0(active0, 0x4000000000000L, active1, 0L);
+         return jjMoveStringLiteralDfa2_0(active0, 0x8000000000000L, active1, 0L);
       case 88:
       case 120:
-         return jjMoveStringLiteralDfa2_0(active0, 0x100000000000L, active1, 0x80L);
+         return jjMoveStringLiteralDfa2_0(active0, 0x200000000000L, active1, 0x100L);
       case 124:
          if ((active0 & 0x80L) != 0L)
             return jjStopAtPos(1, 7);
@@ -548,100 +550,100 @@
    {
       case 65:
       case 97:
-         return jjMoveStringLiteralDfa3_0(active0, 0L, active1, 0x600000L);
+         return jjMoveStringLiteralDfa3_0(active0, 0L, active1, 0xc00000L);
       case 67:
       case 99:
-         if ((active0 & 0x1000000000000L) != 0L)
-            return jjStartNfaWithStates_0(2, 48, 38);
+         if ((active0 & 0x2000000000000L) != 0L)
+            return jjStartNfaWithStates_0(2, 49, 37);
          break;
       case 68:
       case 100:
-         if ((active0 & 0x1000000000L) != 0L)
-            return jjStartNfaWithStates_0(2, 36, 38);
-         else if ((active1 & 0x800L) != 0L)
-            return jjStartNfaWithStates_0(2, 75, 38);
-         return jjMoveStringLiteralDfa3_0(active0, 0x800000000000L, active1, 0x4000L);
+         if ((active0 & 0x2000000000L) != 0L)
+            return jjStartNfaWithStates_0(2, 37, 37);
+         else if ((active1 & 0x1000L) != 0L)
+            return jjStartNfaWithStates_0(2, 76, 37);
+         return jjMoveStringLiteralDfa3_0(active0, 0x1000000000000L, active1, 0x8000L);
       case 69:
       case 101:
-         return jjMoveStringLiteralDfa3_0(active0, 0x8000000800000000L, active1, 0L);
+         return jjMoveStringLiteralDfa3_0(active0, 0x1000000000L, active1, 0x1L);
       case 70:
       case 102:
-         return jjMoveStringLiteralDfa3_0(active0, 0x40000000L, active1, 0L);
+         return jjMoveStringLiteralDfa3_0(active0, 0x80000000L, active1, 0L);
       case 71:
       case 103:
-         if ((active0 & 0x4000000000000L) != 0L)
-            return jjStartNfaWithStates_0(2, 50, 38);
-         else if ((active1 & 0x200L) != 0L)
+         if ((active0 & 0x8000000000000L) != 0L)
+            return jjStartNfaWithStates_0(2, 51, 37);
+         else if ((active1 & 0x400L) != 0L)
          {
-            jjmatchedKind = 73;
+            jjmatchedKind = 74;
             jjmatchedPos = 2;
          }
-         return jjMoveStringLiteralDfa3_0(active0, 0x1000000020000000L, active1, 0x440L);
+         return jjMoveStringLiteralDfa3_0(active0, 0x2000000040000000L, active1, 0x880L);
       case 73:
       case 105:
-         return jjMoveStringLiteralDfa3_0(active0, 0x400100500000000L, active1, 0x100020L);
+         return jjMoveStringLiteralDfa3_0(active0, 0x800200a00000000L, active1, 0x200040L);
       case 75:
       case 107:
-         return jjMoveStringLiteralDfa3_0(active0, 0x40000000000L, active1, 0L);
+         return jjMoveStringLiteralDfa3_0(active0, 0x80000000000L, active1, 0L);
       case 76:
       case 108:
-         return jjMoveStringLiteralDfa3_0(active0, 0x800010080200000L, active1, 0L);
+         return jjMoveStringLiteralDfa3_0(active0, 0x1000020100400000L, active1, 0L);
       case 77:
       case 109:
-         if ((active0 & 0x20000000000000L) != 0L)
-            return jjStartNfaWithStates_0(2, 53, 38);
+         if ((active0 & 0x40000000000000L) != 0L)
+            return jjStartNfaWithStates_0(2, 54, 37);
          break;
       case 78:
       case 110:
-         if ((active0 & 0x10000000000000L) != 0L)
-            return jjStartNfaWithStates_0(2, 52, 38);
-         else if ((active1 & 0x2000000L) != 0L)
-            return jjStartNfaWithStates_0(2, 89, 38);
+         if ((active0 & 0x20000000000000L) != 0L)
+            return jjStartNfaWithStates_0(2, 53, 37);
          else if ((active1 & 0x4000000L) != 0L)
-            return jjStartNfaWithStates_0(2, 90, 38);
-         return jjMoveStringLiteralDfa3_0(active0, 0x2100000008000000L, active1, 0x8000L);
+            return jjStartNfaWithStates_0(2, 90, 37);
+         else if ((active1 & 0x8000000L) != 0L)
+            return jjStartNfaWithStates_0(2, 91, 37);
+         return jjMoveStringLiteralDfa3_0(active0, 0x4200000010000000L, active1, 0x10000L);
       case 79:
       case 111:
-         return jjMoveStringLiteralDfa3_0(active0, 0x200001000000L, active1, 0x80107L);
+         return jjMoveStringLiteralDfa3_0(active0, 0x400002000000L, active1, 0x10020eL);
       case 80:
       case 112:
-         if ((active0 & 0x800000L) != 0L)
-            return jjStartNfaWithStates_0(2, 23, 38);
-         else if ((active1 & 0x80L) != 0L)
-            return jjStartNfaWithStates_0(2, 71, 38);
+         if ((active0 & 0x1000000L) != 0L)
+            return jjStartNfaWithStates_0(2, 24, 37);
+         else if ((active1 & 0x100L) != 0L)
+            return jjStartNfaWithStates_0(2, 72, 37);
          break;
       case 82:
       case 114:
-         return jjMoveStringLiteralDfa3_0(active0, 0x200000000000000L, active1, 0x20000L);
+         return jjMoveStringLiteralDfa3_0(active0, 0x400000000000000L, active1, 0x40000L);
       case 83:
       case 115:
-         if ((active1 & 0x10L) != 0L)
-            return jjStartNfaWithStates_0(2, 68, 38);
-         else if ((active1 & 0x800000L) != 0L)
-            return jjStartNfaWithStates_0(2, 87, 38);
-         return jjMoveStringLiteralDfa3_0(active0, 0x2000000000000L, active1, 0x8L);
+         if ((active1 & 0x20L) != 0L)
+            return jjStartNfaWithStates_0(2, 69, 37);
+         else if ((active1 & 0x1000000L) != 0L)
+            return jjStartNfaWithStates_0(2, 88, 37);
+         return jjMoveStringLiteralDfa3_0(active0, 0x4000000000000L, active1, 0x10L);
       case 84:
       case 116:
-         if ((active0 & 0x4000000000L) != 0L)
-            return jjStartNfaWithStates_0(2, 38, 38);
-         else if ((active1 & 0x1000000L) != 0L)
-            return jjStartNfaWithStates_0(2, 88, 38);
-         return jjMoveStringLiteralDfa3_0(active0, 0x4000020014000000L, active1, 0L);
+         if ((active0 & 0x8000000000L) != 0L)
+            return jjStartNfaWithStates_0(2, 39, 37);
+         else if ((active1 & 0x2000000L) != 0L)
+            return jjStartNfaWithStates_0(2, 89, 37);
+         return jjMoveStringLiteralDfa3_0(active0, 0x8000040028000000L, active1, 0L);
       case 85:
       case 117:
-         return jjMoveStringLiteralDfa3_0(active0, 0x40000000000000L, active1, 0x50000L);
+         return jjMoveStringLiteralDfa3_0(active0, 0x80000000000000L, active1, 0xa0000L);
       case 86:
       case 118:
-         return jjMoveStringLiteralDfa3_0(active0, 0x400000000000L, active1, 0L);
+         return jjMoveStringLiteralDfa3_0(active0, 0x800000000000L, active1, 0L);
       case 87:
       case 119:
-         return jjMoveStringLiteralDfa3_0(active0, 0L, active1, 0x2000L);
+         return jjMoveStringLiteralDfa3_0(active0, 0L, active1, 0x4000L);
       case 88:
       case 120:
-         if ((active0 & 0x8000000000000L) != 0L)
-            return jjStartNfaWithStates_0(2, 51, 38);
-         else if ((active0 & 0x80000000000000L) != 0L)
-            return jjStartNfaWithStates_0(2, 55, 38);
+         if ((active0 & 0x10000000000000L) != 0L)
+            return jjStartNfaWithStates_0(2, 52, 37);
+         else if ((active0 & 0x100000000000000L) != 0L)
+            return jjStartNfaWithStates_0(2, 56, 37);
          break;
       default :
          break;
@@ -660,84 +662,84 @@
    switch(curChar)
    {
       case 49:
-         return jjMoveStringLiteralDfa4_0(active0, 0L, active1, 0x400L);
+         return jjMoveStringLiteralDfa4_0(active0, 0L, active1, 0x800L);
       case 65:
       case 97:
-         if ((active0 & 0x8000000000000000L) != 0L)
-            return jjStartNfaWithStates_0(3, 63, 38);
+         if ((active1 & 0x1L) != 0L)
+            return jjStartNfaWithStates_0(3, 64, 37);
          break;
       case 67:
       case 99:
-         if ((active0 & 0x2000000000000L) != 0L)
-            return jjStartNfaWithStates_0(3, 49, 38);
-         return jjMoveStringLiteralDfa4_0(active0, 0x200000000000000L, active1, 0L);
+         if ((active0 & 0x4000000000000L) != 0L)
+            return jjStartNfaWithStates_0(3, 50, 37);
+         return jjMoveStringLiteralDfa4_0(active0, 0x400000000000000L, active1, 0L);
       case 68:
       case 100:
-         if ((active1 & 0x8000L) != 0L)
-            return jjStartNfaWithStates_0(3, 79, 38);
+         if ((active1 & 0x10000L) != 0L)
+            return jjStartNfaWithStates_0(3, 80, 37);
          break;
       case 69:
       case 101:
-         if ((active0 & 0x40000000000L) != 0L)
-            return jjStartNfaWithStates_0(3, 42, 38);
-         return jjMoveStringLiteralDfa4_0(active0, 0x4000800018200000L, active1, 0x2000L);
+         if ((active0 & 0x80000000000L) != 0L)
+            return jjStartNfaWithStates_0(3, 43, 37);
+         return jjMoveStringLiteralDfa4_0(active0, 0x8001000030400000L, active1, 0x4000L);
       case 72:
       case 104:
-         return jjMoveStringLiteralDfa4_0(active0, 0x20000000L, active1, 0L);
+         return jjMoveStringLiteralDfa4_0(active0, 0x40000000L, active1, 0L);
       case 73:
       case 105:
-         return jjMoveStringLiteralDfa4_0(active0, 0x1000400000000000L, active1, 0x4000L);
+         return jjMoveStringLiteralDfa4_0(active0, 0x2000800000000000L, active1, 0x8000L);
       case 76:
       case 108:
-         if ((active0 & 0x80000000L) != 0L)
-            return jjStartNfaWithStates_0(3, 31, 38);
-         else if ((active0 & 0x10000000000L) != 0L)
-            return jjStartNfaWithStates_0(3, 40, 38);
-         return jjMoveStringLiteralDfa4_0(active0, 0L, active1, 0x20L);
+         if ((active0 & 0x100000000L) != 0L)
+            return jjStartNfaWithStates_0(3, 32, 37);
+         else if ((active0 & 0x20000000000L) != 0L)
+            return jjStartNfaWithStates_0(3, 41, 37);
+         return jjMoveStringLiteralDfa4_0(active0, 0L, active1, 0x40L);
       case 77:
       case 109:
-         if ((active0 & 0x1000000L) != 0L)
-            return jjStartNfaWithStates_0(3, 24, 38);
+         if ((active0 & 0x2000000L) != 0L)
+            return jjStartNfaWithStates_0(3, 25, 37);
          break;
       case 78:
       case 110:
-         if ((active0 & 0x100000000L) != 0L)
-            return jjStartNfaWithStates_0(3, 32, 38);
-         else if ((active1 & 0x100000L) != 0L)
-            return jjStartNfaWithStates_0(3, 84, 38);
+         if ((active0 & 0x200000000L) != 0L)
+            return jjStartNfaWithStates_0(3, 33, 37);
          else if ((active1 & 0x200000L) != 0L)
+            return jjStartNfaWithStates_0(3, 85, 37);
+         else if ((active1 & 0x400000L) != 0L)
          {
-            jjmatchedKind = 85;
+            jjmatchedKind = 86;
             jjmatchedPos = 3;
          }
-         return jjMoveStringLiteralDfa4_0(active0, 0x440000400000000L, active1, 0x450000L);
+         return jjMoveStringLiteralDfa4_0(active0, 0x880000800000000L, active1, 0x8a0000L);
       case 79:
       case 111:
-         return jjMoveStringLiteralDfa4_0(active0, 0L, active1, 0x100L);
+         return jjMoveStringLiteralDfa4_0(active0, 0L, active1, 0x200L);
       case 82:
       case 114:
-         return jjMoveStringLiteralDfa4_0(active0, 0x800000000L, active1, 0x47L);
+         return jjMoveStringLiteralDfa4_0(active0, 0x1000000000L, active1, 0x8eL);
       case 83:
       case 115:
-         if ((active1 & 0x80000L) != 0L)
-            return jjStartNfaWithStates_0(3, 83, 38);
-         return jjMoveStringLiteralDfa4_0(active0, 0x100000000000L, active1, 0L);
+         if ((active1 & 0x100000L) != 0L)
+            return jjStartNfaWithStates_0(3, 84, 37);
+         return jjMoveStringLiteralDfa4_0(active0, 0x200000000000L, active1, 0L);
       case 84:
       case 116:
-         if ((active0 & 0x40000000L) != 0L)
-            return jjStartNfaWithStates_0(3, 30, 38);
-         else if ((active1 & 0x20000L) != 0L)
-            return jjStartNfaWithStates_0(3, 81, 38);
-         return jjMoveStringLiteralDfa4_0(active0, 0x2100000000000000L, active1, 0x8L);
+         if ((active0 & 0x80000000L) != 0L)
+            return jjStartNfaWithStates_0(3, 31, 37);
+         else if ((active1 & 0x40000L) != 0L)
+            return jjStartNfaWithStates_0(3, 82, 37);
+         return jjMoveStringLiteralDfa4_0(active0, 0x4200000000000000L, active1, 0x10L);
       case 85:
       case 117:
-         return jjMoveStringLiteralDfa4_0(active0, 0x200004000000L, active1, 0L);
+         return jjMoveStringLiteralDfa4_0(active0, 0x400008000000L, active1, 0L);
       case 87:
       case 119:
-         return jjMoveStringLiteralDfa4_0(active0, 0x20000000000L, active1, 0L);
+         return jjMoveStringLiteralDfa4_0(active0, 0x40000000000L, active1, 0L);
       case 89:
       case 121:
-         return jjMoveStringLiteralDfa4_0(active0, 0x800000000000000L, active1, 0L);
+         return jjMoveStringLiteralDfa4_0(active0, 0x1000000000000000L, active1, 0L);
       default :
          break;
    }
@@ -755,69 +757,69 @@
    switch(curChar)
    {
       case 48:
-         if ((active1 & 0x400L) != 0L)
-            return jjStartNfaWithStates_0(4, 74, 21);
+         if ((active1 & 0x800L) != 0L)
+            return jjStartNfaWithStates_0(4, 75, 21);
          break;
       case 50:
-         if ((active1 & 0x400000L) != 0L)
-            return jjStartNfaWithStates_0(4, 86, 21);
+         if ((active1 & 0x800000L) != 0L)
+            return jjStartNfaWithStates_0(4, 87, 21);
          break;
       case 65:
       case 97:
-         return jjMoveStringLiteralDfa5_0(active0, 0x2000000000000000L, active1, 0x4008L);
+         return jjMoveStringLiteralDfa5_0(active0, 0x4000000000000000L, active1, 0x8010L);
       case 67:
       case 99:
-         return jjMoveStringLiteralDfa5_0(active0, 0x200000L, active1, 0x40000L);
+         return jjMoveStringLiteralDfa5_0(active0, 0x400000L, active1, 0x80000L);
       case 68:
       case 100:
-         if ((active1 & 0x10000L) != 0L)
-            return jjStartNfaWithStates_0(4, 80, 38);
-         return jjMoveStringLiteralDfa5_0(active0, 0L, active1, 0x7L);
+         if ((active1 & 0x20000L) != 0L)
+            return jjStartNfaWithStates_0(4, 81, 37);
+         return jjMoveStringLiteralDfa5_0(active0, 0L, active1, 0xeL);
       case 69:
       case 101:
-         if ((active0 & 0x800000000L) != 0L)
-            return jjStartNfaWithStates_0(4, 35, 38);
-         return jjMoveStringLiteralDfa5_0(active0, 0x20000000000L, active1, 0x40L);
+         if ((active0 & 0x1000000000L) != 0L)
+            return jjStartNfaWithStates_0(4, 36, 37);
+         return jjMoveStringLiteralDfa5_0(active0, 0x40000000000L, active1, 0x80L);
       case 71:
       case 103:
-         if ((active0 & 0x400000000L) != 0L)
-            return jjStartNfaWithStates_0(4, 34, 38);
-         return jjMoveStringLiteralDfa5_0(active0, 0x800000000000000L, active1, 0L);
+         if ((active0 & 0x800000000L) != 0L)
+            return jjStartNfaWithStates_0(4, 35, 37);
+         return jjMoveStringLiteralDfa5_0(active0, 0x1000000000000000L, active1, 0L);
       case 73:
       case 105:
-         return jjMoveStringLiteralDfa5_0(active0, 0L, active1, 0x20L);
+         return jjMoveStringLiteralDfa5_0(active0, 0L, active1, 0x40L);
       case 76:
       case 108:
-         return jjMoveStringLiteralDfa5_0(active0, 0x200000000000000L, active1, 0L);
+         return jjMoveStringLiteralDfa5_0(active0, 0x400000000000000L, active1, 0L);
       case 78:
       case 110:
-         return jjMoveStringLiteralDfa5_0(active0, 0x400000000000L, active1, 0L);
+         return jjMoveStringLiteralDfa5_0(active0, 0x800000000000L, active1, 0L);
       case 79:
       case 111:
-         return jjMoveStringLiteralDfa5_0(active0, 0x1000000000000000L, active1, 0L);
+         return jjMoveStringLiteralDfa5_0(active0, 0x2000000000000000L, active1, 0L);
       case 80:
       case 112:
-         return jjMoveStringLiteralDfa5_0(active0, 0x200000000000L, active1, 0L);
+         return jjMoveStringLiteralDfa5_0(active0, 0x400000000000L, active1, 0L);
       case 82:
       case 114:
-         if ((active0 & 0x8000000L) != 0L)
-            return jjStartNfaWithStates_0(4, 27, 38);
-         else if ((active0 & 0x10000000L) != 0L)
-            return jjStartNfaWithStates_0(4, 28, 38);
-         else if ((active1 & 0x100L) != 0L)
-            return jjStartNfaWithStates_0(4, 72, 38);
-         else if ((active1 & 0x2000L) != 0L)
-            return jjStartNfaWithStates_0(4, 77, 38);
-         return jjMoveStringLiteralDfa5_0(active0, 0x4100800004000000L, active1, 0L);
+         if ((active0 & 0x10000000L) != 0L)
+            return jjStartNfaWithStates_0(4, 28, 37);
+         else if ((active0 & 0x20000000L) != 0L)
+            return jjStartNfaWithStates_0(4, 29, 37);
+         else if ((active1 & 0x200L) != 0L)
+            return jjStartNfaWithStates_0(4, 73, 37);
+         else if ((active1 & 0x4000L) != 0L)
+            return jjStartNfaWithStates_0(4, 78, 37);
+         return jjMoveStringLiteralDfa5_0(active0, 0x8201000008000000L, active1, 0L);
       case 84:
       case 116:
-         if ((active0 & 0x20000000L) != 0L)
-            return jjStartNfaWithStates_0(4, 29, 38);
-         else if ((active0 & 0x40000000000000L) != 0L)
-            return jjStartNfaWithStates_0(4, 54, 38);
-         else if ((active0 & 0x400000000000000L) != 0L)
-            return jjStartNfaWithStates_0(4, 58, 38);
-         return jjMoveStringLiteralDfa5_0(active0, 0x100000000000L, active1, 0L);
+         if ((active0 & 0x40000000L) != 0L)
+            return jjStartNfaWithStates_0(4, 30, 37);
+         else if ((active0 & 0x80000000000000L) != 0L)
+            return jjStartNfaWithStates_0(4, 55, 37);
+         else if ((active0 & 0x800000000000000L) != 0L)
+            return jjStartNfaWithStates_0(4, 59, 37);
+         return jjMoveStringLiteralDfa5_0(active0, 0x200000000000L, active1, 0L);
       default :
          break;
    }
@@ -835,48 +837,48 @@
    switch(curChar)
    {
       case 32:
-         return jjMoveStringLiteralDfa6_0(active0, 0xa00000000000L, active1, 0L);
+         return jjMoveStringLiteralDfa6_0(active0, 0x1400000000000L, active1, 0L);
       case 49:
-         if ((active1 & 0x1L) != 0L)
-            return jjStartNfaWithStates_0(5, 64, 21);
-         break;
-      case 50:
          if ((active1 & 0x2L) != 0L)
             return jjStartNfaWithStates_0(5, 65, 21);
          break;
+      case 50:
+         if ((active1 & 0x4L) != 0L)
+            return jjStartNfaWithStates_0(5, 66, 21);
+         break;
       case 65:
       case 97:
-         return jjMoveStringLiteralDfa6_0(active0, 0x4000000L, active1, 0x40000L);
+         return jjMoveStringLiteralDfa6_0(active0, 0x8000000L, active1, 0x80000L);
       case 69:
       case 101:
-         if ((active0 & 0x200000000000000L) != 0L)
-            return jjStartNfaWithStates_0(5, 57, 38);
-         return jjMoveStringLiteralDfa6_0(active0, 0x20000000000L, active1, 0x40L);
+         if ((active0 & 0x400000000000000L) != 0L)
+            return jjStartNfaWithStates_0(5, 58, 37);
+         return jjMoveStringLiteralDfa6_0(active0, 0x40000000000L, active1, 0x80L);
       case 71:
       case 103:
-         if ((active0 & 0x400000000000L) != 0L)
-            return jjStartNfaWithStates_0(5, 46, 38);
+         if ((active0 & 0x800000000000L) != 0L)
+            return jjStartNfaWithStates_0(5, 47, 37);
          break;
       case 73:
       case 105:
-         return jjMoveStringLiteralDfa6_0(active0, 0x2000000000000000L, active1, 0L);
+         return jjMoveStringLiteralDfa6_0(active0, 0x4000000000000000L, active1, 0L);
       case 78:
       case 110:
-         if ((active0 & 0x1000000000000000L) != 0L)
-            return jjStartNfaWithStates_0(5, 60, 38);
-         return jjMoveStringLiteralDfa6_0(active0, 0L, active1, 0x4028L);
+         if ((active0 & 0x2000000000000000L) != 0L)
+            return jjStartNfaWithStates_0(5, 61, 37);
+         return jjMoveStringLiteralDfa6_0(active0, 0L, active1, 0x8050L);
       case 79:
       case 111:
-         return jjMoveStringLiteralDfa6_0(active0, 0x900000000000000L, active1, 0L);
+         return jjMoveStringLiteralDfa6_0(active0, 0x1200000000000000L, active1, 0L);
       case 83:
       case 115:
-         if ((active0 & 0x100000000000L) != 0L)
-            return jjStartNfaWithStates_0(5, 44, 38);
-         return jjMoveStringLiteralDfa6_0(active0, 0x4000000000000000L, active1, 0x4L);
+         if ((active0 & 0x200000000000L) != 0L)
+            return jjStartNfaWithStates_0(5, 45, 37);
+         return jjMoveStringLiteralDfa6_0(active0, 0x8000000000000000L, active1, 0x8L);
       case 84:
       case 116:
-         if ((active0 & 0x200000L) != 0L)
-            return jjStartNfaWithStates_0(5, 21, 38);
+         if ((active0 & 0x400000L) != 0L)
+            return jjStartNfaWithStates_0(5, 22, 37);
          break;
       default :
          break;
@@ -896,46 +898,46 @@
    {
       case 66:
       case 98:
-         return jjMoveStringLiteralDfa7_0(active0, 0xa00000000000L, active1, 0L);
+         return jjMoveStringLiteralDfa7_0(active0, 0x1400000000000L, active1, 0L);
       case 67:
       case 99:
-         return jjMoveStringLiteralDfa7_0(active0, 0L, active1, 0x8L);
+         return jjMoveStringLiteralDfa7_0(active0, 0L, active1, 0x10L);
       case 69:
       case 101:
-         return jjMoveStringLiteralDfa7_0(active0, 0x4000000000000000L, active1, 0L);
+         return jjMoveStringLiteralDfa7_0(active0, 0x8000000000000000L, active1, 0L);
       case 71:
       case 103:
-         if ((active1 & 0x20L) != 0L)
-            return jjStartNfaWithStates_0(6, 69, 38);
+         if ((active1 & 0x40L) != 0L)
+            return jjStartNfaWithStates_0(6, 70, 37);
          break;
       case 73:
       case 105:
-         return jjMoveStringLiteralDfa7_0(active0, 0x100000000000000L, active1, 0L);
+         return jjMoveStringLiteralDfa7_0(active0, 0x200000000000000L, active1, 0L);
       case 76:
       case 108:
-         if ((active0 & 0x4000000L) != 0L)
-            return jjStartNfaWithStates_0(6, 26, 38);
+         if ((active0 & 0x8000000L) != 0L)
+            return jjStartNfaWithStates_0(6, 27, 37);
          break;
       case 78:
       case 110:
-         if ((active0 & 0x20000000000L) != 0L)
-            return jjStartNfaWithStates_0(6, 41, 38);
-         else if ((active0 & 0x800000000000000L) != 0L)
-            return jjStartNfaWithStates_0(6, 59, 38);
-         return jjMoveStringLiteralDfa7_0(active0, 0x2000000000000000L, active1, 0L);
+         if ((active0 & 0x40000000000L) != 0L)
+            return jjStartNfaWithStates_0(6, 42, 37);
+         else if ((active0 & 0x1000000000000000L) != 0L)
+            return jjStartNfaWithStates_0(6, 60, 37);
+         return jjMoveStringLiteralDfa7_0(active0, 0x4000000000000000L, active1, 0L);
       case 83:
       case 115:
-         if ((active1 & 0x40L) != 0L)
-            return jjStartNfaWithStates_0(6, 70, 38);
-         else if ((active1 & 0x4000L) != 0L)
-            return jjStartNfaWithStates_0(6, 78, 38);
+         if ((active1 & 0x80L) != 0L)
+            return jjStartNfaWithStates_0(6, 71, 37);
+         else if ((active1 & 0x8000L) != 0L)
+            return jjStartNfaWithStates_0(6, 79, 37);
          break;
       case 84:
       case 116:
-         return jjMoveStringLiteralDfa7_0(active0, 0L, active1, 0x40000L);
+         return jjMoveStringLiteralDfa7_0(active0, 0L, active1, 0x80000L);
       case 89:
       case 121:
-         return jjMoveStringLiteralDfa7_0(active0, 0L, active1, 0x4L);
+         return jjMoveStringLiteralDfa7_0(active0, 0L, active1, 0x8L);
       default :
          break;
    }
@@ -954,32 +956,32 @@
    {
       case 67:
       case 99:
-         return jjMoveStringLiteralDfa8_0(active0, 0x4000000000000000L, active1, 0L);
+         return jjMoveStringLiteralDfa8_0(active0, 0x8000000000000000L, active1, 0L);
       case 68:
       case 100:
-         if ((active0 & 0x100000000000000L) != 0L)
-            return jjStartNfaWithStates_0(7, 56, 38);
+         if ((active0 & 0x200000000000000L) != 0L)
+            return jjStartNfaWithStates_0(7, 57, 37);
          break;
       case 69:
       case 101:
-         if ((active1 & 0x8L) != 0L)
-            return jjStartNfaWithStates_0(7, 67, 38);
-         else if ((active1 & 0x40000L) != 0L)
-            return jjStartNfaWithStates_0(7, 82, 38);
+         if ((active1 & 0x10L) != 0L)
+            return jjStartNfaWithStates_0(7, 68, 37);
+         else if ((active1 & 0x80000L) != 0L)
+            return jjStartNfaWithStates_0(7, 83, 37);
          break;
       case 83:
       case 115:
-         if ((active0 & 0x2000000000000000L) != 0L)
-            return jjStartNfaWithStates_0(7, 61, 38);
-         else if ((active1 & 0x4L) != 0L)
-            return jjStartNfaWithStates_0(7, 66, 38);
+         if ((active0 & 0x4000000000000000L) != 0L)
+            return jjStartNfaWithStates_0(7, 62, 37);
+         else if ((active1 & 0x8L) != 0L)
+            return jjStartNfaWithStates_0(7, 67, 37);
          break;
       case 89:
       case 121:
-         if ((active0 & 0x200000000000L) != 0L)
-            return jjStopAtPos(7, 45);
-         else if ((active0 & 0x800000000000L) != 0L)
-            return jjStopAtPos(7, 47);
+         if ((active0 & 0x400000000000L) != 0L)
+            return jjStopAtPos(7, 46);
+         else if ((active0 & 0x1000000000000L) != 0L)
+            return jjStopAtPos(7, 48);
          break;
       default :
          break;
@@ -999,7 +1001,7 @@
    {
       case 84:
       case 116:
-         return jjMoveStringLiteralDfa9_0(active0, 0x4000000000000000L);
+         return jjMoveStringLiteralDfa9_0(active0, 0x8000000000000000L);
       default :
          break;
    }
@@ -1018,8 +1020,8 @@
    {
       case 83:
       case 115:
-         if ((active0 & 0x4000000000000000L) != 0L)
-            return jjStartNfaWithStates_0(9, 62, 38);
+         if ((active0 & 0x8000000000000000L) != 0L)
+            return jjStartNfaWithStates_0(9, 63, 37);
          break;
       default :
          break;
@@ -1056,32 +1058,32 @@
                case 21:
                   if ((0x3ff000000000000L & l) == 0L)
                      break;
-                  if (kind > 100)
-                     kind = 100;
+                  if (kind > 101)
+                     kind = 101;
                   jjCheckNAdd(21);
                   break;
                case 38:
-                  if ((0x3ff000000000000L & l) == 0L)
-                     break;
-                  if (kind > 100)
-                     kind = 100;
-                  jjCheckNAdd(21);
-                  break;
-               case 37:
                   if ((0x3ff000000000000L & l) != 0L)
                   {
-                     if (kind > 103)
-                        kind = 103;
+                     if (kind > 104)
+                        kind = 104;
                      jjCheckNAdd(27);
                   }
                   if ((0x3ff000000000000L & l) != 0L)
                      jjCheckNAddTwoStates(23, 24);
                   break;
+               case 37:
+                  if ((0x3ff000000000000L & l) == 0L)
+                     break;
+                  if (kind > 101)
+                     kind = 101;
+                  jjCheckNAdd(21);
+                  break;
                case 0:
                   if ((0x3ff000000000000L & l) != 0L)
                   {
-                     if (kind > 104)
-                        kind = 104;
+                     if (kind > 105)
+                        kind = 105;
                      jjCheckNAddStates(0, 6);
                   }
                   else if ((0x100002600L & l) != 0L)
@@ -1103,29 +1105,29 @@
                case 11:
                   if ((0x3ff000000000000L & l) == 0L)
                      break;
-                  if (kind > 100)
-                     kind = 100;
+                  if (kind > 101)
+                     kind = 101;
                   jjCheckNAdd(21);
                   break;
                case 16:
                   if ((0x3ff000000000000L & l) == 0L)
                      break;
-                  if (kind > 100)
-                     kind = 100;
+                  if (kind > 101)
+                     kind = 101;
                   jjCheckNAdd(21);
                   break;
                case 13:
                   if ((0x3ff000000000000L & l) == 0L)
                      break;
-                  if (kind > 100)
-                     kind = 100;
+                  if (kind > 101)
+                     kind = 101;
                   jjCheckNAdd(21);
                   break;
                case 10:
                   if ((0x3ff000000000000L & l) == 0L)
                      break;
-                  if (kind > 100)
-                     kind = 100;
+                  if (kind > 101)
+                     kind = 101;
                   jjCheckNAdd(21);
                   break;
                case 1:
@@ -1138,7 +1140,7 @@
                   break;
                case 3:
                   if (curChar == 62)
-                     kind = 16;
+                     kind = 17;
                   break;
                case 4:
                   if (curChar == 60)
@@ -1146,7 +1148,7 @@
                   break;
                case 5:
                   if (curChar == 61)
-                     kind = 16;
+                     kind = 17;
                   break;
                case 6:
                   if (curChar == 33)
@@ -1159,8 +1161,8 @@
                case 19:
                   if (curChar != 45)
                      break;
-                  if (kind > 91)
-                     kind = 91;
+                  if (kind > 92)
+                     kind = 92;
                   jjCheckNAdd(19);
                   break;
                case 22:
@@ -1178,22 +1180,22 @@
                case 26:
                   if ((0x3ff000000000000L & l) == 0L)
                      break;
-                  if (kind > 102)
-                     kind = 102;
+                  if (kind > 103)
+                     kind = 103;
                   jjCheckNAdd(26);
                   break;
                case 27:
                   if ((0x3ff000000000000L & l) == 0L)
                      break;
-                  if (kind > 103)
-                     kind = 103;
+                  if (kind > 104)
+                     kind = 104;
                   jjCheckNAdd(27);
                   break;
                case 28:
                   if ((0x3ff000000000000L & l) == 0L)
                      break;
-                  if (kind > 104)
-                     kind = 104;
+                  if (kind > 105)
+                     kind = 105;
                   jjCheckNAddStates(0, 6);
                   break;
                case 29:
@@ -1219,22 +1221,22 @@
                case 34:
                   if (curChar != 46)
                      break;
-                  if (kind > 103)
-                     kind = 103;
+                  if (kind > 104)
+                     kind = 104;
                   jjCheckNAdd(35);
                   break;
                case 35:
                   if ((0x3ff000000000000L & l) == 0L)
                      break;
-                  if (kind > 103)
-                     kind = 103;
+                  if (kind > 104)
+                     kind = 104;
                   jjCheckNAdd(35);
                   break;
                case 36:
                   if ((0x3ff000000000000L & l) == 0L)
                      break;
-                  if (kind > 104)
-                     kind = 104;
+                  if (kind > 105)
+                     kind = 105;
                   jjCheckNAdd(36);
                   break;
                default : break;
@@ -1251,38 +1253,38 @@
                case 12:
                   if ((0x7fffffe87fffffeL & l) != 0L)
                   {
-                     if (kind > 100)
-                        kind = 100;
+                     if (kind > 101)
+                        kind = 101;
                      jjCheckNAdd(21);
                   }
                   if ((0x7fffffe07fffffeL & l) != 0L)
                   {
-                     if (kind > 100)
-                        kind = 100;
+                     if (kind > 101)
+                        kind = 101;
                      jjCheckNAddTwoStates(20, 21);
                   }
                   if ((0x8000000080000L & l) != 0L)
                      jjstateSet[jjnewStateCnt++] = 11;
                   break;
-               case 38:
+               case 37:
                   if ((0x7fffffe87fffffeL & l) != 0L)
                   {
-                     if (kind > 100)
-                        kind = 100;
+                     if (kind > 101)
+                        kind = 101;
                      jjCheckNAdd(21);
                   }
                   if ((0x7fffffe07fffffeL & l) != 0L)
                   {
-                     if (kind > 100)
-                        kind = 100;
+                     if (kind > 101)
+                        kind = 101;
                      jjCheckNAddTwoStates(20, 21);
                   }
                   break;
                case 0:
                   if ((0x7fffffe07fffffeL & l) != 0L)
                   {
-                     if (kind > 100)
-                        kind = 100;
+                     if (kind > 101)
+                        kind = 101;
                      jjCheckNAddTwoStates(20, 21);
                   }
                   if ((0x200000002L & l) != 0L)
@@ -1293,14 +1295,14 @@
                case 11:
                   if ((0x7fffffe87fffffeL & l) != 0L)
                   {
-                     if (kind > 100)
-                        kind = 100;
+                     if (kind > 101)
+                        kind = 101;
                      jjCheckNAdd(21);
                   }
                   if ((0x7fffffe07fffffeL & l) != 0L)
                   {
-                     if (kind > 100)
-                        kind = 100;
+                     if (kind > 101)
+                        kind = 101;
                      jjCheckNAddTwoStates(20, 21);
                   }
                   if ((0x10000000100000L & l) != 0L)
@@ -1309,14 +1311,14 @@
                case 16:
                   if ((0x7fffffe87fffffeL & l) != 0L)
                   {
-                     if (kind > 100)
-                        kind = 100;
+                     if (kind > 101)
+                        kind = 101;
                      jjCheckNAdd(21);
                   }
                   if ((0x7fffffe07fffffeL & l) != 0L)
                   {
-                     if (kind > 100)
-                        kind = 100;
+                     if (kind > 101)
+                        kind = 101;
                      jjCheckNAddTwoStates(20, 21);
                   }
                   if ((0x100000001000L & l) != 0L)
@@ -1325,14 +1327,14 @@
                case 13:
                   if ((0x7fffffe87fffffeL & l) != 0L)
                   {
-                     if (kind > 100)
-                        kind = 100;
+                     if (kind > 101)
+                        kind = 101;
                      jjCheckNAdd(21);
                   }
                   if ((0x7fffffe07fffffeL & l) != 0L)
                   {
-                     if (kind > 100)
-                        kind = 100;
+                     if (kind > 101)
+                        kind = 101;
                      jjCheckNAddTwoStates(20, 21);
                   }
                   if ((0x20000000200L & l) != 0L)
@@ -1341,22 +1343,22 @@
                case 10:
                   if ((0x7fffffe87fffffeL & l) != 0L)
                   {
-                     if (kind > 100)
-                        kind = 100;
+                     if (kind > 101)
+                        kind = 101;
                      jjCheckNAdd(21);
                   }
                   if ((0x7fffffe07fffffeL & l) != 0L)
                   {
-                     if (kind > 100)
-                        kind = 100;
+                     if (kind > 101)
+                        kind = 101;
                      jjCheckNAddTwoStates(20, 21);
                   }
                   if ((0x20000000200L & l) != 0L)
                      jjstateSet[jjnewStateCnt++] = 9;
                   break;
                case 7:
-                  if ((0x10000000100000L & l) != 0L && kind > 22)
-                     kind = 22;
+                  if ((0x10000000100000L & l) != 0L && kind > 23)
+                     kind = 23;
                   break;
                case 8:
                   if ((0x800000008L & l) != 0L)
@@ -1371,8 +1373,8 @@
                      jjstateSet[jjnewStateCnt++] = 13;
                   break;
                case 15:
-                  if ((0x100000001000L & l) != 0L && kind > 22)
-                     kind = 22;
+                  if ((0x100000001000L & l) != 0L && kind > 23)
+                     kind = 23;
                   break;
                case 17:
                   if ((0x200000002L & l) != 0L)
@@ -1381,15 +1383,15 @@
                case 20:
                   if ((0x7fffffe07fffffeL & l) == 0L)
                      break;
-                  if (kind > 100)
-                     kind = 100;
+                  if (kind > 101)
+                     kind = 101;
                   jjCheckNAddTwoStates(20, 21);
                   break;
                case 21:
                   if ((0x7fffffe87fffffeL & l) == 0L)
                      break;
-                  if (kind > 100)
-                     kind = 100;
+                  if (kind > 101)
+                     kind = 101;
                   jjCheckNAdd(21);
                   break;
                case 24:
@@ -1442,7 +1444,7 @@
    switch(curChar)
    {
       case 39:
-         return jjStartNfaWithStates_2(0, 96, 1);
+         return jjStartNfaWithStates_2(0, 97, 1);
       default :
          return jjMoveNfa_2(0, 0);
    }
@@ -1476,15 +1478,15 @@
                case 0:
                   if ((0xffffff7fffffffffL & l) != 0L)
                   {
-                     if (kind > 95)
-                        kind = 95;
+                     if (kind > 96)
+                        kind = 96;
                   }
                   else if (curChar == 39)
                      jjstateSet[jjnewStateCnt++] = 1;
                   break;
                case 1:
-                  if (curChar == 39 && kind > 95)
-                     kind = 95;
+                  if (curChar == 39 && kind > 96)
+                     kind = 96;
                   break;
                case 2:
                   if (curChar == 39)
@@ -1502,7 +1504,7 @@
             switch(jjstateSet[--i])
             {
                case 0:
-                  kind = 95;
+                  kind = 96;
                   break;
                default : break;
             }
@@ -1517,8 +1519,8 @@
             switch(jjstateSet[--i])
             {
                case 0:
-                  if ((jjbitVec0[i2] & l2) != 0L && kind > 95)
-                     kind = 95;
+                  if ((jjbitVec0[i2] & l2) != 0L && kind > 96)
+                     kind = 96;
                   break;
                default : break;
             }
@@ -1562,15 +1564,15 @@
                case 0:
                   if ((0x2400L & l) != 0L)
                   {
-                     if (kind > 92)
-                        kind = 92;
+                     if (kind > 93)
+                        kind = 93;
                   }
                   if (curChar == 13)
                      jjstateSet[jjnewStateCnt++] = 1;
                   break;
                case 1:
-                  if (curChar == 10 && kind > 92)
-                     kind = 92;
+                  if (curChar == 10 && kind > 93)
+                     kind = 93;
                   break;
                case 2:
                   if (curChar == 13)
@@ -1623,13 +1625,14 @@
 /** Token literal values. */
 public static final String[] jjstrLiteralImages = {
 "", null, "\50", "\51", "\56", "\54", "\73", "\174\174", "\53", "\55", "\52", 
-"\57", "\46", "\174", "\136", "\75", null, "\74", "\74\75", "\76", "\76\75", null, 
+"\57", "\45", "\46", "\174", "\136", "\75", null, "\74", "\74\75", "\76", "\76\75", 
 null, null, null, null, null, null, null, null, null, null, null, null, null, null, 
 null, null, null, null, null, null, null, null, null, null, null, null, null, null, 
 null, null, null, null, null, null, null, null, null, null, null, null, null, null, 
 null, null, null, null, null, null, null, null, null, null, null, null, null, null, 
 null, null, null, null, null, null, null, null, null, null, null, null, null, null, 
-null, null, null, null, null, null, null, null, null, null, null, null, null, null, };
+null, null, null, null, null, null, null, null, null, null, null, null, null, null, 
+null, };
 
 /** Lexer state names. */
 public static final String[] lexStateNames = {
@@ -1644,17 +1647,17 @@
    -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, 
    -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, 
    -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, 
-   -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, 1, 0, -1, 2, -1, 0, 3, -1, 0, 
-   -1, -1, -1, -1, -1, -1, 
+   -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, 1, 0, -1, 2, -1, 0, 3, -1, 
+   0, -1, -1, -1, -1, -1, -1, 
 };
 static final long[] jjtoToken = {
-   0xfffffffffffffffdL, 0x1d907ffffffL, 
+   0xfffffffffffffffdL, 0x3b20fffffffL, 
 };
 static final long[] jjtoSkip = {
-   0x2L, 0x10000000L, 
+   0x2L, 0x20000000L, 
 };
 static final long[] jjtoMore = {
-   0x0L, 0x6e8000000L, 
+   0x0L, 0xdd0000000L, 
 };
 protected SimpleCharStream input_stream;
 private final int[] jjrounds = new int[37];
@@ -1769,9 +1772,9 @@
          jjmatchedKind = 0x7fffffff;
          jjmatchedPos = 0;
          curPos = jjMoveStringLiteralDfa0_1();
-         if (jjmatchedPos == 0 && jjmatchedKind > 93)
+         if (jjmatchedPos == 0 && jjmatchedKind > 94)
          {
-            jjmatchedKind = 93;
+            jjmatchedKind = 94;
          }
          break;
        case 2:
diff -r 65b854cbf712 -r 5e5206c8b3fc adql/parser/adqlGrammar.jj
--- a/adql/parser/adqlGrammar.jj	Wed Dec 03 20:54:03 2014 +0200
+++ b/adql/parser/adqlGrammar.jj	Wed Dec 03 22:07:16 2014 +0200
@@ -506,6 +506,7 @@
 |	< MINUS: "-" >
 |	< ASTERISK: "*" >
 |	< DIVIDE: "/" >
+|	< MODULO: "%" >
 }
 
 /* ******************** */
diff -r 65b854cbf712 -r 5e5206c8b3fc adql/query/operand/OperationType.java
--- a/adql/query/operand/OperationType.java	Wed Dec 03 20:54:03 2014 +0200
+++ b/adql/query/operand/OperationType.java	Wed Dec 03 22:07:16 2014 +0200
@@ -28,10 +28,10 @@
  * @see Operation
  */
 public enum OperationType{
-	SUM, SUB, MULT, DIV;
+	SUM, SUB, MULT, DIV, MOD, BIT_OR, BIT_AND, BIT_XOR;
 
 	public static String[] getOperators(){
-		return new String[]{SUM.toString(),SUB.toString(),MULT.toString(),DIV.toString()};
+		return new String[]{SUM.toString(),SUB.toString(),MULT.toString(),DIV.toString(), MOD.toString(), BIT_OR.toString(), BIT_AND.toString(), BIT_XOR.toString() };
 	}
 
 	public static OperationType getOperator(String str) throws UnsupportedOperationException{
@@ -43,6 +43,14 @@
 			return MULT;
 		else if (str.equalsIgnoreCase("/"))
 			return DIV;
+	    else if (str.equalsIgnoreCase("%"))
+            return MOD;
+        else if (str.equalsIgnoreCase("|"))
+            return BIT_OR;
+        else if (str.equalsIgnoreCase("&"))
+            return BIT_AND;
+        else if (str.equalsIgnoreCase("^"))
+            return BIT_XOR;
 		else
 			throw new UnsupportedOperationException("Numeric operation unknown: \"" + str + "\" !");
 	}
@@ -62,6 +70,14 @@
 				return "*";
 			case DIV:
 				return "/";
+		   case MOD:
+	            return "%";
+	        case BIT_OR:
+	            return "|";
+	        case BIT_AND:
+	            return "&";
+	        case BIT_XOR:
+	            return "^";
 			default:
 				return "???";
 		}
